### PR TITLE
Context menu rework

### PR DESCRIPTION
**Pull request by gera-gas**
Resolves: #998

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] I made sure the code compiles on my machine
- [x] I made sure there are no unnecessary changes in the code
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving
- [x] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"

### CODE REVIEW

**Comment by gera-gas** [comment]
Pending of #1258 for add "Rename clip" menu item.

File: src/TrackPanel.cpp [hunk 1]
```diff
@@ -735,11 +742,62 @@ void TrackPanel::OnMouseEvent(wxMouseEvent & event)
          }
       } );
    }
+    
+    if(event.RightDown())
```
Comment (Paul-Licameli) [inline src/TrackPanel.cpp:739]: Context menu should happen at mouse up, not down, and on Windows should also happen for the special menu key.  See how LabelTrackView does it.

File: src/TrackPanel.cpp [hunk 2]
```diff
@@ -735,11 +742,62 @@ void TrackPanel::OnMouseEvent(wxMouseEvent & event)
...
+            contextMenu.Bind(wxEVT_MENU,
+               [this]( wxCommandEvent &event ){
+                  OnContextMenu( /*GetProject(),*/ event ); }
+            );
+            
+            contextMenu.Append(kContextMenuItemID_AddMonoTrack, XO("Add mono track").Translation());
+            contextMenu.Append(kContextMenuItemID_AddStereoTrack, XO("Add stereo track").Translation());
+            contextMenu.Append(kContextMenuItemID_AddLabelTrack, XO("Add label track").Translation());
+            contextMenu.AppendSeparator();
+            contextMenu.Append(kContextMenuItemID_ExportProject, XO("Export project\tCtrl+E").Translation());
```
Comment (Paul-Licameli) [inline src/TrackPanel.cpp]: You have a hard-coded menu shortcut, which we now know is wrong.  You correct this in a later commit.  I don't want PRs past the draft stage that still keep a history of errors and corrections.

File: src/TrackPanel.cpp [hunk 3]
```diff
@@ -735,11 +742,62 @@ void TrackPanel::OnMouseEvent(wxMouseEvent & event)
...
+    }
 
    // Must also fall through to base class handler
    event.Skip();
 }
 
+#include "SelectUtilities.h"
+
+void TrackPanel::OnContextMenu(/*AudacityProject &project, */ wxCommandEvent& event)
+{
```
Comment (Paul-Licameli) [inline src/TrackPanel.cpp:782]: I don't like making TrackPanel the class that handles this.  Maybe it should be done within class BackgroundCell instead, by analogy with how LabelTrackView handles context menus.

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -652,6 +652,7 @@ list( APPEND SOURCES
       menus/SelectMenus.cpp
       menus/ToolbarMenus.cpp
       menus/TrackMenus.cpp
+      menus/TrackMenus.h
```
Comment (Paul-Licameli) [inline src/CMakeLists.txt:657]: No.  Do not create this header.

File: src/TrackPanel.cpp [hunk 5]
```diff
@@ -774,25 +776,35 @@ void TrackPanel::OnMouseEvent(wxMouseEvent & event)
...
+#include "menus/TrackMenus.h"
 
 void TrackPanel::OnContextMenu(/*AudacityProject &project, */ wxCommandEvent& event)
 {
     switch(event.GetId())
     {
         case kContextMenuItemID_AddMonoTrack:
+            this->CallAfter([this]{
+                TrackActions::Handler().OnNewWaveTrack(CommandContext{*GetProject()});
+            });
```
Comment (Paul-Licameli) [inline src/TrackPanel.cpp:788]: No.  Do not dispatch the commands by direct linkage to what used to be private functions.

I ask that it be done instead by lookup of a textual name in CommandManager::HandleTextualCommand.

File: src/TrackPanel.cpp [hunk 6]
```diff
@@ -754,18 +760,36 @@ void TrackPanel::OnMouseEvent(wxMouseEvent & event)
...
-            contextMenu.Append(kContextMenuItemID_AddStereoTrack, XO("Add stereo track").Translation());
-            contextMenu.Append(kContextMenuItemID_AddLabelTrack, XO("Add label track").Translation());
-            contextMenu.AppendSeparator();
-            contextMenu.Append(kContextMenuItemID_ExportProject, XO("Export project\tCtrl+E").Translation());
-            contextMenu.AppendSeparator();
-            contextMenu.Append(kContextMenuItemID_SelectAll, XO("Select All\tCtrl+A").Translation());
+            // Get cursor position area: on track or non track (empty area).
+            const auto track = FindTrack (FindCell(event.m_x, event.m_y).pCell.get());
+            if(track)
+            {
```
Comment (Paul-Licameli) [inline src/TrackPanel.cpp:756]: Did you try it with a Label track?  Those tracks already have their own context menus.

I think these context menus should appear only on Wave tracks.

File: src/TrackPanel.cpp [hunk 7]
```diff
@@ -754,18 +760,36 @@ void TrackPanel::OnMouseEvent(wxMouseEvent & event)
...
+                contextMenu.Append(kContextMenuItemID_Cut, XO("Cut\tCtrl+X").Translation());
+                contextMenu.Append(kContextMenuItemID_Copy, XO("Copy\tCtrl+C").Translation());
```
Comment (Paul-Licameli) [inline src/TrackPanel.cpp]: I don't like these repetitious calls to Append.  The description of the menu should be declarative.  There should just be a table of command names, for lookup in CommandManager, and user-visible names.  Use something special like null for the separators.  Then write one simple loop that iterates the table, and make menu items or separators.

File: src/TrackPanel.cpp [hunk 8]
```diff
@@ -810,6 +834,30 @@ void TrackPanel::OnContextMenu(/*AudacityProject &project, */ wxCommandEvent& ev
     }
 }
 
+void TrackPanel::OnTrackAreaContextMenu(wxCommandEvent& event)
+{
+    switch(event.GetId())
+    {
```
Comment (Paul-Licameli) [inline src/TrackPanel.cpp:818]: And I don't like this switch statement.  Instead you should be able to get an ID from `event`, index into the table that describes the menu, and invoke CommandManager::HandleTextualCommand.

File: src/CMakeLists.txt [hunk 9]
```diff
@@ -643,6 +643,7 @@ list( APPEND SOURCES
 
       menus/ClipMenus.cpp
       menus/EditMenus.cpp
+      menus/EditMenus.h
```
Comment (Paul-Licameli) [inline src/CMakeLists.txt:646]: No again to this new header file.

File: src/TrackPanel.h [hunk 10]
```diff
@@ -224,4 +231,109 @@ struct AUDACITY_DLL_API IsVisibleTrack
    wxRect mPanelRect;
 };
 
+//! A class with context menu over audio track.
+class AUDACITY_DLL_API AudacityOverTrackContextMenu : public wxMenu
```
Comment (Paul-Licameli) [inline src/TrackPanel.h:236]: Putting all of this into a header file instead of a .cpp file is unnecessary.

File: src/TrackPanel.h [hunk 11]
```diff
@@ -224,4 +231,109 @@ struct AUDACITY_DLL_API IsVisibleTrack
...
+            cmdName = ComponentInterfaceSymbol(wxT("MuteTracks"), muteItemString);
+            muteItemString = commandManager->DescribeCommandsAndShortcuts(&cmdName, 1u);
+        }
+        
+        Append(MenuItemID::kItemID_Cut, cutItemString.Translation());
+        Append(MenuItemID::kItemID_Copy, copyItemString.Translation());
+        Append(MenuItemID::kItemID_Paste, pasteItemString.Translation());
+        AppendSeparator();
+        Append(MenuItemID::kItemID_Split, splitItemString.Translation());
+        Append(MenuItemID::kItemID_Mute, muteItemString.Translation());
```
Comment (Paul-Licameli) [inline src/TrackPanel.h:285]: Again, make it table-driven.  Not repetitions calls to Append().

Comment (Paul-Licameli) [reply src/TrackPanel.h:285]: Another problem is that the event dispatching code must specially cooperate with this menu construction code.  Much simpler just to have a table-driven way to do both things, and one common table top describe it.

File: src/TrackPanel.h [hunk 12]
```diff
@@ -35,6 +35,7 @@
 #include "SelectUtilities.h"
 #include "menus/TrackMenus.h"
 #include "menus/EditMenus.h"
+#include "menus/FileMenus.h"
```
Comment (Paul-Licameli) [inline src/TrackPanel.h:38]: Do not make this new header file.

**Review by Paul-Licameli — CHANGES_REQUESTED** [review]
Too much is wrong with this PR.

A better approach would be table-driven.  A declarative, not procedural, description of the menus, and small "interpreters" to build the menus and dispatch the events.

**Pull request closed without merging**